src/main/java/com/microsoft/azure/relay/HybridConnectionListener.java

```diff
@@ -538,26 +536,22 @@ public void setOnlineHandler(BiConsumer<Object, Object[]> onOnline) {
 		 *         is established between the control websocket and the cloud service
 		 */
 		public CompletableFuture<Void> openAsync(Duration timeout) {
-			// Establish a WebSocket connection right now so we can detect any fatal errors
-			return CompletableFuture.runAsync(() -> {
-				CompletableFuture<Void> connectTask = null;
-				boolean success = false;
-				
-				try {
-					// Block so we surface any errors to the user right away.
-					connectTask = this.ensureConnectTask(timeout);
-					connectTask.join();
-
+			// Establish a WebSocket connection right now so we can detect any fatal errors		
+			CompletableFuture<ClientWebSocket> connectTask = this.ensureConnectTask(timeout);
+			return connectTask.thenApply((webSocket) -> {
 					this.tokenRenewer.setOnTokenRenewed((token) -> this.onTokenRenewed(token));
-					this.receivePumpAsync();
-					success = true;
-				} finally {
-					if (!success) {
-						this.closeOrAbortWebSocketAsync(connectTask, new CloseReason(CloseCodes.UNEXPECTED_CONDITION,
-								"closing web socket connection because something went wrong trying to connect."));			
+					this.receivePumpAsync(); // This starts pumping but doesn't wait
+					return (Void)null;
```

Comment:
>(Void) [](start = 12, length = 6)

Is this cast needed?

Reply:
It wouldn't compile without either this cast, specifying `.<Void>.thenApply` or putting it into a local variable of type `CompletableFuture<Void>`.

---
In reply to: [258629544](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258629544) [](ancestors = 258629544)

Reply:
That's an indication of maybe .thenApply isn't the best fit here. Will .thenAccept work?

---
In reply to: [258631053](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258631053) [](ancestors = 258631053,258629544)

Reply:
Ooooooh. Awesome info for me to learn! I'll try it out.

---
In reply to: [258632412](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258632412) [](ancestors = 258632412,258631053,258629544)

Reply:
Fixed in iteration 3 (iteration 2 was a merge with master).

---
In reply to: [258633157](https://github.com/Azure/azure-relay-java/pull/17#discussion_r258633157) [](ancestors = 258633157,258632412,258631053,258629544)